solvers/H3LAPD/Diagnostics/GrowthRatesRecorder.hpp

```diff
@@ -21,138 +21,159 @@ namespace NESO::Solvers::H3LAPD {
-    Vmath::Vvtvp(m_npts, zderiv, 1, zderiv, 1, integrand, 1, integrand, 1);
-
-    // integrand *= 0.5
-    Vmath::Smul(m_npts, 0.5, integrand, 1, integrand, 1);
-    return m_n->Integral(integrand);
+    Vmath::Vmul(this->npts, this->n->GetPhys(), 1, this->n->GetPhys(), 1,
+                integrand, 1);
+
+    // Compute ϕ derivs, square them and add to integrand
+    Array<OneD, NekDouble> xderiv(this->npts), yderiv(this->npts),
+        zderiv(this->npts);
+    this->phi->PhysDeriv(this->phi->GetPhys(), xderiv, yderiv, zderiv);
+    Vmath::Vvtvp(this->npts, xderiv, 1, xderiv, 1, integrand, 1, integrand, 1);
+    Vmath::Vvtvp(this->npts, yderiv, 1, yderiv, 1, integrand, 1, integrand, 1);
+    if (this->prob_ndims == 2) {
+      /* Should be ∇⊥ϕ^2, so ∂ϕ/∂z ought to be excluded in both 2D and
+       * 3D, but there's a small discrepancy in 2D without it. Energy 'leaking'
+       * into orthogonal dimension?!? */
+      Vmath::Vvtvp(this->npts, zderiv, 1, zderiv, 1, integrand, 1, integrand,
+                   1);
+    }
+
+    return 0.5 * this->n->Integral(integrand);
   }
 
   /**
    * Calculate Enstrophy = 0.5 ∫ (n-w)^2 dV
    */
   inline double compute_enstrophy() {
-    Array<OneD, NekDouble> integrand(m_npts);
+    Array<OneD, NekDouble> integrand(this->npts);
     // Set integrand = n-w
-    Vmath::Vsub(m_npts, m_n->GetPhys(), 1, m_w->GetPhys(), 1, integrand, 1);
+    Vmath::Vsub(this->npts, this->n->GetPhys(), 1, this->w->GetPhys(), 1,
+                integrand, 1);
     // Set integrand = (n-w)^2
-    Vmath::Vmul(m_npts, integrand, 1, integrand, 1, integrand, 1);
-    // Set integrand = 0.5*(n-w)^2
-    Vmath::Smul(m_npts, 0.5, integrand, 1, integrand, 1);
-    return m_n->Integral(integrand);
+    Vmath::Vmul(this->npts, integrand, 1, integrand, 1, integrand, 1);
+    return 0.5 * this->n->Integral(integrand);
   }
 
   /**
-   * Calculate Gamma_alpha = alpha ∫ (n-phi)^2 dV
+   * Calculate Γα
+   *  In 2D: Γα = α ∫ (n-ϕ)^2 dV​
+   *  In 3D: Γα = α ∫ [∂/∂z(n-ϕ)]^2 dV
    */
   inline double compute_Gamma_a() {
-    Array<OneD, NekDouble> integrand(m_npts);
-    // Set integrand = n - phi
-    Vmath::Vsub(m_npts, m_n->GetPhys(), 1, m_phi->GetPhys(), 1, integrand, 1);
-    // Set integrand = (n - phi)^2
-    Vmath::Vmul(m_npts, integrand, 1, integrand, 1, integrand, 1);
-    // Set integrand = alpha*(n - phi)^2
-    Vmath::Smul(m_npts, m_alpha, integrand, 1, integrand, 1);
-    return m_n->Integral(integrand);
+    Array<OneD, NekDouble> integrand(this->npts);
+    // Set integrand = n - ϕ
+    Vmath::Vsub(this->npts, this->n->GetPhys(), 1, this->phi->GetPhys(), 1,
+                integrand, 1);
+
+    switch (this->prob_ndims) {
+    case 2:
+      // Set integrand = (n - ϕ)^2
+      Vmath::Vmul(this->npts, integrand, 1, integrand, 1, integrand, 1);
+      break;
+    case 3:
+      // Set integrand = ∂/∂z(n-ϕ)
+      this->phi->PhysDeriv(2, integrand, integrand);
```

Comment:
For functions like `PhysDeriv` (and possibly Vmul etc?) I would be careful aliasing the inputs and outputs. Here inarray is a const array and out_d is not a const array hence using the same array for both arguments could cause undefined behavior (?).

[1] https://doc.nektar.info/doxygen/latest/class_nektar_1_1_multi_regions_1_1_exp_list.html#ae1afc120a27ecfa1fb37d692264e37d2

Reply:
Thanks, good point.
I've modified GrowthRatesRecorder to use temporary arrays instead for all Vmath and PhysDeriv calls.
Also created [this issue](https://github.com/ExCALIBUR-NEPTUNE/NESO/issues/278) to check the rest of the code - I've almost certainly used this shortcut elsewhere.